torch/_functorch/aot_autograd.py

```diff
@@ -3247,7 +3247,20 @@ def runtime_wrapper(*args):
                     if trace_joint:
                         assert isinstance(updated_inpt, TensorAlias)
                         updated_inpt = updated_inpt.alias
-                    original_inpt.set_(updated_inpt)
+                    try:
+                        original_inpt.set_(updated_inpt)
+                    except RuntimeError:
```

Comment:
Maybe this is fine as an unblock, just pasting some thoughts below:

---

It feels like the right thing to do is to store metadata at trace time that tells us when the `_unsafe_set_version_counter` and `no_grad` are necessary to use at runtime.

`_unsafe_set_version_counter`:
Check if each input's VC changed during tracing, and only use `_unsafe_set_version_counter` if the VC didn't change during tracing. It's probably fine to only track this for inputs marked with `mutates_storage_metadata`. This will also not be good enough, though, if we have an input that got a `.data=` mutation, **and** had another mutation from a normal operator that bumped its version counter (we'll incorrectly bump the VC a second time at runtime). This case is... extremely unlikely though?

`no_grad`:
Option 1 (this might be enough for the FSDP case): only run under `no_grad()` if `original_inpt.is_leaf`. We know that the `set_()` would have raised an error without no_grad in this case, and we should have already error'd out at compile time if the user tried to do this. We should add a torch.compile test for this error case though if we do this check

Option 2: Every time we trace a set_() call, test whether or not happened under no_grad at trace time, so we can pass that info along to runtime. One way is in the set_() implementation for functionalization, check `is_grad_enabled()` and store that info on the `FunctionalTensorWrapper`. This also won't be enough, though, if we e.g. have an input that we called `set_()` on multiple times, sometimes with grad enabled and sometimes with grad disabled.

Reply:
Let me think about this some more - it's a good thought. The idea of storing metadata at trace time to guide counter lowering and no_grad usage is a good one.